flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/source/enumerator/ContinuousSplitPlannerImpl.java

```diff
@@ -0,0 +1,170 @@
+    table.refresh();
+    if (lastPosition != null) {
+      return discoverDeltaSplits(lastPosition);
+    } else {
+      return discoverInitialSplits();
+    }
+  }
+
+  /**
+   * Discover delta changes between @{code lastPosition} and current table snapshot
+   */
+  private ContinuousEnumerationResult discoverDeltaSplits(IcebergEnumeratorPosition lastPosition) {
+    // incremental discovery mode
+    Snapshot currentSnapshot = table.currentSnapshot();
+    if (currentSnapshot.snapshotId() == lastPosition.endSnapshotId()) {
+      LOG.info("Current table snapshot is already enumerated: {}", currentSnapshot.snapshotId());
+      return new ContinuousEnumerationResult(Collections.emptyList(), lastPosition);
+    } else {
+      ScanContext incrementalScan = scanContext
+          .copyWithAppendsBetween(lastPosition.endSnapshotId(), currentSnapshot.snapshotId());
+      LOG.info("Incremental scan: startSnapshotId = {}, endSnapshotId = {}",
+          incrementalScan.startSnapshotId(), incrementalScan.endSnapshotId());
+      List<IcebergSourceSplit> splits = FlinkSplitPlanner.planIcebergSourceSplits(table, incrementalScan, workerPool);
+      IcebergEnumeratorPosition position = IcebergEnumeratorPosition.builder()
+          .startSnapshotId(lastPosition.endSnapshotId())
+          .startSnapshotTimestampMs(lastPosition.endSnapshotTimestampMs())
+          .endSnapshotId(currentSnapshot.snapshotId())
+          .endSnapshotTimestampMs(currentSnapshot.timestampMillis())
+          .build();
+      LOG.info("Discovered {} splits from incremental scan: {}", splits.size(), position);
+      return new ContinuousEnumerationResult(splits, position);
+    }
+  }
+
+  /**
+   * Discovery initial set of splits based on {@link StreamingStartingStrategy}
+   */
+  private ContinuousEnumerationResult discoverInitialSplits() {
+    HistoryEntry startSnapshotEntry = getStartSnapshot(table, scanContext);
+    LOG.info("get startSnapshotId {} based on starting strategy {}",
+        startSnapshotEntry.snapshotId(), scanContext.startingStrategy());
+    List<IcebergSourceSplit> splits;
+    if (scanContext.startingStrategy() ==
+        StreamingStartingStrategy.TABLE_SCAN_THEN_INCREMENTAL) {
+      // do a full table scan first
+      splits = FlinkSplitPlanner.planIcebergSourceSplits(table, scanContext, workerPool);
+      LOG.info("Discovered {} splits from initial full table scan with snapshotId {}",
+          splits.size(), startSnapshotEntry);
+    } else {
+      splits = Collections.emptyList();
+    }
+
+    IcebergEnumeratorPosition position = IcebergEnumeratorPosition.builder()
+        .endSnapshotId(startSnapshotEntry.snapshotId())
+        .endSnapshotTimestampMs(startSnapshotEntry.timestampMillis())
+        .build();
+    return new ContinuousEnumerationResult(splits, position);
+  }
+
+  @VisibleForTesting
+  static HistoryEntry getStartSnapshot(Table table, ScanContext scanContext) {
```

Comment:
Here you probably want to use `SnapshotUtil`, which handles most of these cases using the current table state's ancestors.

Reply:
will switch